File: src/application/controllers/NavigationController.ts
import { Controller } from "../../core/controllers/Controller";
import { Property } from "../../core/observable/property/Property";
import { GuiStore, GuiTool } from "../../core/stores/GuiStore";
import { property } from "../../core/observable";
import { Clock } from "../../core/Clock";

export class NavigationController extends Controller {
    private readonly _internet_time = property("@");
    private readonly internet_time_interval: any;

    readonly tool: Property<GuiTool>;
    readonly internet_time: Property<string> = this._internet_time;

    constructor(private readonly gui_store: GuiStore, private readonly clock: Clock) {
        super();

        this.tool = gui_store.tool;
        this.internet_time_interval = setInterval(this.set_internet_time, 1000);
        this.set_internet_time();
    }

    dispose(): void {
        super.dispose();
        clearInterval(this.internet_time_interval);
    }

    set_tool(tool: GuiTool): void {
        this.gui_store.set_tool(tool);
    }

    private set_internet_time = (): void => {
        const now = this.clock.now();
        const s = now.getUTCSeconds();
        const m = now.getUTCMinutes();
        const h = (now.getUTCHours() + 1) % 24; // Internet time is calculated from UTC+01:00.
        this._internet_time.val = `@${Math.floor((s + 60 * (m + 60 * h)) / 86.4)}`;
    };
}

File: web/src/main/kotlin/world/phantasmal/web/application/controllers/NavigationController.kt
package world.phantasmal.web.application.controllers

import kotlinx.browser.window
import kotlinx.datetime.Clock
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import world.phantasmal.cell.Cell
import world.phantasmal.cell.mutableCell
import world.phantasmal.web.core.PwToolType
import world.phantasmal.web.core.stores.UiStore
import world.phantasmal.webui.controllers.Controller
import kotlin.math.floor

class NavigationController(private val uiStore: UiStore, private val clock: Clock) : Controller() {
    private val _internetTime = mutableCell("@")
    private var internetTimeInterval: Int

    val tools: Map<PwToolType, Cell<Boolean>> = uiStore.toolToActive
    val internetTime: Cell<String> = _internetTime

    init {
        internetTimeInterval = window.setInterval(::updateInternetTime, 1000)
        updateInternetTime()
    }

    override fun dispose() {
        window.clearInterval(internetTimeInterval)
        super.dispose()
    }

    fun setCurrentTool(tool: PwToolType) {
        uiStore.setCurrentTool(tool)
    }

    private fun updateInternetTime() {
        val now = clock.now().toLocalDateTime(INTERNET_TIME_TZ)
        _internetTime.value = "@" + floor((now.second + 60 * (now.minute + 60 * now.hour)) / 86.4)
    }

    companion object {
        /**
         * Internet time is calculated from UTC+01:00.
         */
        private val INTERNET_TIME_TZ = TimeZone.of("UTC+01:00")
    }
}
